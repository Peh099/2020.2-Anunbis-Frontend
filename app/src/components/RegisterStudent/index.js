import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { useHistory } from 'react-router-dom';
import schema from './validations';
import Form from '../Form';
import Input from '../Input';
import Select from '../Select';
import Button from '../Button';
import api from '../../services/Api';
import { getCourses } from '../../services/Courses';

export default function RegisterStudent({ onRegister }) {
  const history = useHistory();
  const [courses, setCourses] = React.useState([]);
  const [errorDB, setErrorDB] = React.useState('');
  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(schema),
  });

  function coursesOptions(coursesO) {
    const coursesArray = [];
    coursesO?.map((course) =>
      coursesArray.push({ id: course.id_course, name: course.name }),
    );
    return coursesArray;
  }

  useEffect(() => {
    getCourses(setCourses);
  }, []);

  function onSubmit(data) {
    const body = {
      reg_student: Number(data.reg_student),
      name: data.name,
      id_course: Number(data.id_course),
      email: data.email,
      password: data.password,
    };

    api
      .post('/student', body)
      .then((response) => {
        if (response.status === 201) {
          onRegister('Confirme o seu cadastro com o e-mail de verificação enviado ao seu e-mail.');
        }
      })
      .catch((error) => {
        if (error.response.status === 409) {
          setErrorDB('Estudante já cadastrado');
        }
      });
  }
  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Form.Field errorMsg={errors.name?.message}>
        <Input
          type="text"
          text="Nome"
          name="name"
          register={register}
          data-testid="input-name"
        />
      </Form.Field>
      <Form.Field>
        <Select
          id="courses"
          backColor="#FFD54F"
          text="Selecione um Curso"
          options={coursesOptions(courses)} /* eslint-disable-line */
          name="id_course"
          register={register}
          error={errors.id_course !== undefined}
          data-testid="select-courses"
        />
      </Form.Field>
      <Form.Field errorMsg={errors.reg_student?.message}>
        <Input
          type="text"
          text="Matrícula"
          name="reg_student"
          register={register}
          data-testid="input-reg"
        />
      </Form.Field>
      <Form.Field errorMsg={errors.email?.message}>
        <Input
          type="email"
          text="Email Institucional"
          name="email"
          register={register}
          data-testid="input-email"
        />
      </Form.Field>
      <Form.Field errorMsg={errors.password?.message}>
        <Input
          type="password"
          text="Senha"
          name="password"
          register={register}
          data-testid="input-password"
        />
      </Form.Field>
      <Form.Field errorMsg={errors.co_password?.message}>
        <Input
          type="password"
          text="Confirmar Senha"
          name="co_password"
          register={register}
          data-testid="input-co-password"
        />
      </Form.Field>
      <Form.Field>
        <div className="errorDB">{errorDB}</div>
      </Form.Field>
      <Form.Footer>
        <Button
          text="CANCELAR"
          backColor="#FFF9C4"
          padding="12px 8px"
          onClick={() => history?.push('/')}
        />
        <Button
          text="CONFIRMAR"
          type="submit"
          backColor="#FFF9C4"
          padding="12px 8px"
        />
      </Form.Footer>
    </Form>
  );
}
