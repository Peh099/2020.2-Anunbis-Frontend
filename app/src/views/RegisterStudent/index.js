import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import { Link, useHistory } from 'react-router-dom';
import schema from './validations';
import Form from '../../components/Form';
import Input from '../../components/Input';
import Select from '../../components/Select';
import Button from '../../components/Button';
import api from '../../services/Api';
import {
  Conteiner, Content, Links, Title,
} from './styles';

const Header = ({ children, title }) => (
  <Conteiner>
    <Links>
      {children}
    </Links>
    <Title>{title}</Title>
  </Conteiner>
);

export default function RegisterStudent() {
  const history = useHistory();
  const [courses, setCourses] = React.useState([]);
  const [errorDB, setErrorDB] = React.useState('');
  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(schema),
  });

  function coursesOptions(coursesO) {
    const coursesArray = [];
    coursesO?.map((course) => coursesArray.push({ id: course.id_course, name: course.name }));
    return (
      coursesArray
    );
  }

  useEffect(() => {
    api.get('/course')
      .then((response) => {
        setCourses(response.data);
      });
  }, []);

  function onSubmit(data) {
    const body = {
      reg_student: Number(data.reg_student),
      name: data.name,
      id_course: Number(data.id_course),
      email: data.email,
      password: data.password,
    };

    api.post('/student', body)
      .then((response) => {
        if (response.status === 201) {
          history.push('/visitant/login');
        }
      })
      .catch((error) => {
        if (error.response.status === 409) {
          setErrorDB('Estudante já cadastrado');
        }
      });
  }

  return (
    <Content>
      <Header title="Cadastro de Aluno">
        <Link className="btnLogin" to="/visitant/login">LOGIN</Link>
        <Link className="btnCadastro" to="/visitant/student">CADASTRO</Link>
        <Link className="btnProfessor" to="/visitant/professor">PROFESSOR?</Link>
      </Header>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Field errorMsg={errors.name?.message}><Input type="text" text="Nome" name="name" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.id_course?.message}><Select id="courses" backColor="#FFD54F" text="Selecione um Curso" options={coursesOptions(courses)} name="id_course" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.reg_student?.message}><Input type="text" text="Matrícula" name="reg_student" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.email?.message}><Input type="email" text="Email Institucional" name="email" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.password?.message}><Input type="password" text="Senha" name="password" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.co_password?.message}><Input type="password" text="Confirmar Senha" name="co_password" register={register} /></Form.Field>
        <Form.Field><div className="errorDB">{errorDB}</div></Form.Field>
        <Form.Footer>
          <Button text="CANCELAR" backColor="#FFF9C4" padding="12px 8px" onClick={() => history.push('/')} />
          <Button text="CONFIRMAR" type="submit" backColor="#FFF9C4" padding="12px 8px" />
        </Form.Footer>
      </Form>
    </Content>
  );
}
