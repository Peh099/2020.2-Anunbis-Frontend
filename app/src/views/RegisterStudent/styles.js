import styled from 'styled-components';

export const Conteiner = styled.div`
  width: 400px;
`;

export const Content = styled.div`
  height: 450px;
  width: 400px;
  display:flex;
  align-items: center;
  flex-direction: column;

Form {
  height: 300px;
  width: 300px;
  Input, Select{
      margin-bottom: 5px;
      width: 180px;
    }
  
}

select {
  margin: 0px;
  width: 185px;
}
`;

export const Links = styled.div`
  width: 400px;
  display:flex;
  align-items: center;
  justify-content: space-evenly;
  font-size:14px;
  margin-top: 15px;


  .btnCadastro {
    margin-right: 70px;
    color: #212121;
  }

  .btnLogin, .btnProfessor{
    text-decoration: none;
    color: #212121;
  }
`;

export const Title = styled.h4`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom:20px;
  margin-top: 40px;
`;