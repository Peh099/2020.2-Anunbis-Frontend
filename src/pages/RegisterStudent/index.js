import  React from 'react';
import BaseAutentication from '../../components/BaseAutentication';
import './index.css'


export default function RegisterStudent(){
    return (
      <div className="RegisterStudent">
        <header className="App-header">
          <BaseAutentication />
        </header>
      </div>
    );
}
