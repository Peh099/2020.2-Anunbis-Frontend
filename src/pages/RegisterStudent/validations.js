import * as yup from 'yup';

const schema = yup.object().shape({
    name: yup
    .string()
    .min(2, "Nome deve conter pelo menos dois caracteres.")
    .required("Nome deve ser preenchido."),

    id_course: yup
    .string()
    .required("Curso deve ser preenchido."),

    reg_student: yup
    .string()
    .length(9)
    .matches(/^[0-9]*$/, "Matricula deve conter somente números.")
    .required("Matricula deve ser preenchida."),

    email: yup
    .string()
    .matches(/^[0-9]+@aluno.unb.br$/, "Formato do email deve ser: \"matricula@aluno.unb.br\"")
    .required("Email deve ser preenchido."),

    password: yup
    .string()
    .min(8, "Senha deve ter pelo menos 8 caracteres.")
    .required("Senha deve ser preenchida."),

    co_password: yup
    .string()
    .oneOf([yup.ref('password')], "Senhas devem ser iguais.")
    .required("Confirmação da senha deve ser preenchida."),
});

export default schema;