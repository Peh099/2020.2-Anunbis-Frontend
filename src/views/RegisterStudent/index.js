import React, { useEffect } from 'react';
import styled from 'styled-components';
import { useForm } from "react-hook-form";
import { yupResolver } from '@hookform/resolvers/yup';
import { Link, useHistory } from 'react-router-dom';
import schema from './validations';
import Form from '../../components/Form';
import Input from '../../components/Input';
import Select from '../../components/Select';
import Button from '../../components/Button';

const Conteiner = styled.div`
  width: 400px;
`;

const Links = styled.div`
  width: 400px;
  display:flex;
  align-items: center;
  justify-content: space-evenly;
  font-size:14px;
  margin-top: 15px;


  .btnCadastro {
    margin-right: 70px;
    color: #212121;
  }

  .btnLogin, .btnProfessor{
    text-decoration: none;
    color: #212121;
  }
`;

const Title = styled.h4`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom:20px;
  margin-top: 40px;
`;

const Header = ({ children, title }) => {
  return (
    <Conteiner>
      <Links>
        {children}
      </Links>
      <Title>{title}</Title>
    </Conteiner>
  )
}

const Content = styled.div`
  height: 450px;
  width: 400px;
  display:flex;
  align-items: center;
  flex-direction: column;

Form {
  height: 300px;
  width: 300px;
  Input, Select{
    margin-bottom: 10px;
    width: 180px;
  }
  
}

select {
  margin: 0px;
  width: 185px;
}
`;

export default function RegisterStudent() {
  const history = useHistory();
  const [courses, setCourses] = React.useState([])
  const [errorDB, setErrorDB] = React.useState('')
  const { register, handleSubmit, errors } = useForm({
    resolver: yupResolver(schema),
  });

  console.log(process.env)

  function courses_Options(courses) {
    const coursesArray = [{}]
    courses?.map((course) => coursesArray.push({ id: course.id_course, name: course.name }))
    return (
      coursesArray
    );
  }

  useEffect(() => {
    async function fetchData() {
      const url = process.env.REACT_APP_API_HOST + "/course";
      const response = await fetch(url)
      const data = await response.json()

      setCourses(data);
    }
    fetchData();
  }, []);

  function onSubmit(data) {
    const url = process.env.REACT_APP_API_HOST + "/student";
    const body = {
      reg_student: parseInt(data.reg_student),
      name: data.name,
      id_course: parseInt(data.id_course),
      email: data.email,
      password: data.password
    }

    fetch(url, {
      method: 'post',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body)
    })
      .then(response => response)
      .then(rs => {
        console.log(rs)
        console.log(rs.json())
        if (rs.ok) {
          setErrorDB("")
          history.push("/login")
        }
        if (rs.status === 409) {
          setErrorDB("Estudante já cadastrado")
        }
      })
  };

  return (
    <Content>
      <Header title="Cadastro de Aluno">
        <Link className="btnLogin" to="/user/login">LOGIN</Link>
        <Link className="btnCadastro" to="/user/student">CADASTRO</Link>
        <Link className="btnProfessor" to="/user/professor">PROFESSOR?</Link>
      </Header>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Field errorMsg={errors.name?.message}><Input type="text" text="Nome" name="name" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.id_course?.message}><Select id="courses" options={courses_Options(courses)} name="id_course" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.reg_student?.message}><Input type="text" text="Matrícula" name="reg_student" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.email?.message}><Input type="email" text="Email Institucional" name="email" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.password?.message}><Input type="password" text="Senha" name="password" register={register} /></Form.Field>
        <Form.Field errorMsg={errors.co_password?.message}><Input type="password" text="Confirmar Senha" name="co_password" register={register} /></Form.Field>
        <Form.Field><div className="errorDB">{errorDB}</div></Form.Field>
        <Form.Footer>
          <Button text="CANCELAR" backColor="#FFF9C4" padding="12px 8px" />
          <Button text="CONFIRMAR" type="submit" backColor="#FFF9C4" padding="12px 8px" />
        </Form.Footer>
      </Form>
    </Content>
  )
}