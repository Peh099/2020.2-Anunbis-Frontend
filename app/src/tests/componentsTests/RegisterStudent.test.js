import renderer from 'react-test-renderer';
import RegisterStudent from '../../components/RegisterStudent';
import { screen, render, waitFor } from '@testing-library/react';
import userEvent from "@testing-library/user-event";


// const validStudent = {
//     name: 'Estudante teste',
//     reg: '123456789',
//     email: '123456789@aluno.unb.br',
//     password: '12345678'
// }

describe('Snapshot RegisterStudent component', () => {
    it('matches the snapshot', () => {
        const tree = renderer.create(<RegisterStudent />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})

describe('Testing forms functionability', () => {

    it('window location must be change when click on CANCELAR', () => {
        render(<RegisterStudent />);
        const buttonCancel = screen.getByText('CANCELAR');
        userEvent.click(buttonCancel);

        expect(window.location.pathname).toEqual('/');
    })

    // it('Submit Valid Student', async () => {
    //     const courses = document.createElement("option");
    //     courses.id_course = "1";
    //     courses.value = "1";
    //     courses.text = 'Engenharia de Software';

    //     const callback = jest.fn();
    //     render(<RegisterStudent redirect={callback} />)

    //     const inputName = screen.getByTestId('input-name')
    //     const selectCourses = screen.getByTestId('select-courses')
    //     const inputReg = screen.getByTestId('input-reg')
    //     const inputEmail = screen.getByTestId('input-email')
    //     const inputPassword = screen.getByTestId('input-password')
    //     const inputCoPassword = screen.getByTestId('input-co-password')
    //     const btnConfirm = screen.getByText('CONFIRMAR')

    //     getCourses()

    //     selectCourses.add(courses, null)
    //     console.log("teste stetar teste " + selectCourses.options.length)
    //     userEvent.type(inputName, validStudent.name)
    //     userEvent.selectOptions(selectCourses, ['1'])
    //     userEvent.type(inputReg, validStudent.reg)
    //     userEvent.type(inputEmail, validStudent.email)
    //     userEvent.type(inputPassword, validStudent.password)
    //     userEvent.type(inputCoPassword, validStudent.password)
    //     userEvent.click(btnConfirm);

    //     //expect(screen.getByRole('option', { name: 'Engenharia de Software' }).selected).toBe(true)

    //     await waitFor(() => {
    //         expect(callback).not.toHaveBeenCalled();
    //     })

    // })
})
