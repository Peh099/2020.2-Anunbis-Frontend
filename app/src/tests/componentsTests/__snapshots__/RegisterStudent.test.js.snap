// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot RegisterStudent component matches the snapshot 1`] = `
<form
  className="sc-dlnjwi iDwcvG"
  onSubmit={[Function]}
>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <input
      className="sc-eCApnc doXpOj"
      data-testid="input-name"
      name="name"
      placeholder="Nome"
      type="text"
    />
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <select
      className="sc-jSFjdj bURWPa"
      data-testid="select-courses"
      id="courses"
      name="id_course"
    >
      <option
        disabled={true}
        selected={true}
        value=""
      >
        Selecione um Curso
      </option>
    </select>
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <input
      className="sc-eCApnc doXpOj"
      data-testid="input-reg"
      name="reg_student"
      placeholder="MatrÃ­cula"
      type="text"
    />
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <input
      className="sc-eCApnc doXpOj"
      data-testid="input-email"
      name="email"
      placeholder="Email Institucional"
      type="email"
    />
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <input
      className="sc-eCApnc doXpOj"
      data-testid="input-password"
      name="password"
      placeholder="Senha"
      type="password"
    />
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <input
      className="sc-eCApnc doXpOj"
      data-testid="input-co-password"
      name="co_password"
      placeholder="Confirmar Senha"
      type="password"
    />
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-bdnxRM fcTkcg"
  >
    <div
      className="errorDB"
    >
      
    </div>
    <div
      className="sc-gtsrHT iKVqJa error"
    />
  </div>
  <div
    className="sc-hKFxyN teyFV"
  >
    <button
      className="sc-gKAaRy fldcqT"
      onClick={[Function]}
    >
      CANCELAR
    </button>
    <button
      className="sc-gKAaRy fldcqT"
      type="submit"
    >
      CONFIRMAR
    </button>
  </div>
</form>
`;
